### PR TITLE
Restaurant website

### LINKED ISSUE #42: Create templates to render in cases where restaurant isn't returned
- [ ] Assign 400 errors code and templates to render throughout code as necessary. 
- [ ] Include more sad path testing and verify in SwaggerUI

Identified Issues: 
Restaurant is not returned during certain hours of the day because we are searching with open_now params. 

Proposed Solutions:
1. Remove open_now params and display is_closed (already being returned) and operating hours, so app can be used any time of the day and user can decide for themselves. 
2. Return message that no restaurant is open. 

Implemented Solution:
Returned message that no restaurant is open. App should be used when user is about to eat and user would be angry to receive restaurant that may not be open. 

Identified Issue: 
Restaurant not being returned within certain radius. 

Proposed Solutions:
1. Rerun search without radius param to ensure user is returned a restaurant.
2. Rerun the search only if user selected driving and return error message if user selected walking.
3. User should be notified that no restaurants match their search and be given opportunity to re-enter their preferences. 

Implemented Solution:
Rerun the search only if user selected driving and return error message if user selected walking. Similar reasons for solution as first issue above.

### LINKED ISSUE #46: EXTENSION: Retrieve restaurant link
**Background:** Yelp uses Grubhub to supply links to a restaurant's website and menu. Grubhub is not a public API. 

**Possibilities:**

1. The /search endpoint on the Yelp API has a link available but it is to the Yelp page. If included, this would require the user to click to the Yelp page and find the link on the Yelp page to be able to access it. _This is an alternative solution because not as user friendly as the one provided by using the Google Maps API._
2. Google Maps/Places has a direct link to the restaurant (if provided) and can be obtained with the approach below. This way, a link that goes right to the restaurant's website can be available on the results page of our app. _We have decided to go this route._

**Steps to Get Website Link Using Google API:**
1. Use the find place from text end point at _maps.googleapis.com/maps/api/place/findplacefromtext/json?_ endpoint with the following params (key, locationbias, input, and inputtype) to obtain a place_id (this is required for the next endpoint).
2. Use the details endpoint (a place_id is required) at  _maps.googleapis.com/maps/api/place/details/json?_ with the following params (key and place_id).
3. If a website is returned, it will be at the end of the response under "website". 
4. If there is no website provided, we need to have FE hide that detail from the user on the results page of our application. 
5. This URL should be returned with the rest of the restaurant information we are returning to FE. 

**Links:** 
https://developers.google.com/places/web-service/search
https://developers.google.com/places/web-service/details

**Images from Scott's Walkthrough:**
Screen Shot 2020-04-14 at 9 51 02 AM
Screen Shot 2020-04-14 at 9 52 16 AM
Screen Shot 2020-04-14 at 9 51 57 AM

### PR DESCRIPTION
**Pull request by Jonpatt92**
## What functionality does this accomplish?
closes #42
closes #46

**Description:**

- Extends Google Service to retrieve the Restaurant's Website from the Google Places API and display it in the JSON response of the `/recommendations` endpoint. 
Will default to the Restaurant's Yelp Business API if no valid coordinates are obtained for the Restaurant or if the Google Places API response is not usable. 

- Added error handling throughout Yelp Service for the edge case of no open Restaurants being available. Service will now try to remove 'radius' parameters (IF the User hasn't selected Walking as their Travel mode) and recursively re-call the request with the updated parameters. If that new request fails, it will return an error saying there are no open restaurants near the User. 

 
## What did you struggle on to complete?
N/A

### Test Coverage Percentage: x%
- [ ] No Tests have been changed
- [x] New Tests have been implemented
- [x] Some Tests have been changed
- [ ] All of the Tests have been changed(Please describe what in the world happened):

## Checklist:
- [x] My code has no unused/commented out code
- [x] I have reviewed my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have fully tested my code
- [x] I have partially tested my code (please explain why): Need a way to mock data for various error handling since requests are sensitive to Restaurant's hours of operation

## Helpful Resources:
* Resource link AND small description of info gathered/learned

## Review Requests(optional):
@linda-le1 
@freeheeling 

## Please include a gif of how you feel about this branch:
alt-text

### CODE REVIEW

File: tests/yelp_service_test.py [hunk 1]
```diff
@@ -33,6 +33,9 @@ def test_can_get_recommendation():
     assert 'photos' in response['data']['attributes']
     assert response['data']['attributes']['is_closed'] == False
 
+def test_can_tell_when_no_nearby_restaurants():
+    params = {'latitude': '90.00', 'longitude': '135.00'}
```
Comment (freeheeling) [inline tests/yelp_service_test.py:37]: Where in the world?

Comment (Jonpatt92) [reply tests/yelp_service_test.py:37]: Santa Clause ;)

File: base_app/services/yelp_service.py [hunk 2]
```diff
@@ -4,33 +4,49 @@
...
-        return result
+        if restaurant_id:
+            url = 'https://api.yelp.com/v3/businesses/{}'.format(restaurant_id)
+            response = self.connection(url)
+            json_data = json.dumps(response.json())
+            if 'coordinates' in response.json().keys():
+                google_service = GoogleService()
+                website = google_service.get_website(json_data)
+            elif 'url' in response.json().keys(): # Will return the Restaurant's Yelp Business Page if no Coordinates available
+                website = response.json()['url']
```
Comment (freeheeling) [inline base_app/services/yelp_service.py:23]: Thoughtful provisional.

File: base_app/services/google_service.py [hunk 3]
```diff
@@ -24,4 +23,26 @@ def get_coordinates(self, params):
             content = {'Invalid Address': 'Please enter a street address, city, and state.'}
             return content
 
+    def get_place_id(self, params):
+        url = 'https://maps.googleapis.com/maps/api/place/findplacefromtext/json'
+        params.update({'inputtype': 'textquery'})
+        response = self.connection(url, params)
+        json = response.json()
+        if response.status_code == 200 and 'candidates' in json.keys():
```
Comment (freeheeling) [inline base_app/services/google_service.py]: Do both conditions have to be met? Is it not sufficient to only check if 'candidates' is a key?

Comment (Jonpatt92) [reply base_app/services/google_service.py]: I'm following Linda's lead here. However I could see how there might be a KeyError by trying to run `json.keys()` if `json` ended up being a null variable

File: base_app/services/google_service.py [hunk 4]
```diff
@@ -24,4 +23,26 @@ def get_coordinates(self, params):
...
+        place_id = self.get_place_id(place_id_params)
+
+        params = {'place_id': place_id}
+        url = 'https://maps.googleapis.com/maps/api/place/details/json'
+        response = self.connection(url, params)
+        json_result = response.json()
+        if response.status_code == 200 and 'website' in json_result.keys():
+            return json_result['result']['website']
+        else:
+            return yelp_data['url']
```
Comment (freeheeling) [inline base_app/services/google_service.py]: Viable alternative.

File: base_app/services/yelp_service.py [hunk 5]
```diff
@@ -4,33 +4,49 @@
...
+            recommendation = Restaurant.from_json(json_data, website)
+            schema = RestaurantSchema()
+            result = schema.dump(recommendation)
+            return result
+        elif 'radius' in params.keys() and params['radius'] == 16000:
+            params.pop('radius', None)
+            self.get_recommendation(params)
```
Comment (freeheeling) [inline base_app/services/yelp_service.py:32]: Does this reference the walk option?

File: base_app/services/yelp_service.py [hunk 6]
```diff
@@ -4,33 +4,49 @@
...
-        return restaurant
+        if response.status_code == 200 and 'businesses' in all.keys():
+            for x in all['businesses']:
+                if x.get('price') != None:
+                    ids.append(x['id'])
+                else:
+                    continue
+        if len(ids) != 0:
+            restaurant_id = random.choice(ids)
+            return restaurant_id
```
Comment (freeheeling) [inline base_app/services/yelp_service.py:49]: Thank you for making this more explicit.

**Review by freeheeling — APPROVED** [review]
Great work incorporating this new website addition. Thanks for taking the time to answer the questions I posed.

File: base_app/services/yelp_service.py (hunk 5 above)
Comment (Jonpatt92) [reply base_app/services/yelp_service.py:32]: Line 30 does. The walk option will set the radius to 1,600. If it's set to 16000 then it's either default or Drive.

**Pull request merged**